### PR TITLE
make possible to set the auth token from an env variable

### PR DESCRIPTION
**Pull request by EPedrotti**
**Proposed changes**:
- This is needed to run smoothly on openshift. Right now on docker-compose the starting command for nlu gets interpolated and the token is correctly set. This does not happen in openshift becouse the command is run outside docker-compose and the env variable are not available for interpolations in the starting command but are available to the application. 
 
**Status (please check what you already did)**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

File: rasa_nlu/server.py [hunk 1]
```diff
@@ -406,6 +408,25 @@ def unload_model(self, request):
...
+    clitoken = cmdline_args.token
+    envtoken = os.environ["RASA_NLU_TOKEN"]
+    if conflicting_tokens(clitoken, envtoken):
+        raise Exception(
+            "RASA_NLU_TOKEN is set both with the -t option,"
+            " with value `{}`, and with and environment variable, "
+            "with value `{}`. "
+            "Please set the token with just one method "
+            "to avoid unexpected behaviours.".format(
+                clitoken, envtoken))
```
Comment (wochinge) [inline rasa_nlu/server.py]: can we maybe use the command line token by default and only log a warning when the environment token is overwritten?

File: rasa_nlu/server.py [hunk 2]
```diff
@@ -406,6 +408,25 @@ def unload_model(self, request):
...
 
 
+def conflicting_tokens(token1, token2):
+    return token1 is not None and token2 is not None
+
+
+def get_token():
+    clitoken = cmdline_args.token
+    envtoken = os.environ["RASA_NLU_TOKEN"]
+    if conflicting_tokens(clitoken, envtoken):
```
Comment (wochinge) [inline rasa_nlu/server.py]:
```suggestion
    if clitoken and envtoken:
```

File: rasa_nlu/server.py [hunk 3]
```diff
@@ -406,6 +408,25 @@ def unload_model(self, request):
             return simplejson.dumps({"error": "{}".format(e)})
 
 
+def conflicting_tokens(token1, token2):
+    return token1 is not None and token2 is not None
+
+
+def get_token():
+    clitoken = cmdline_args.token
```
Comment (wochinge) [inline rasa_nlu/server.py]: might be worth adding underscore to the variable names, e.g. `cli_token`

File: rasa_nlu/server.py [hunk 4]
```diff
@@ -429,13 +450,12 @@ def unload_model(self, request):
...
             pre_load = router.project_store.keys()
         router._pre_load(pre_load)
-
     rasa = RasaNLU(
         router,
         cmdline_args.loglevel,
         cmdline_args.write,
         cmdline_args.num_threads,
-        cmdline_args.token,
+        get_token(),
```
Comment (wochinge) [inline rasa_nlu/server.py]: I would change that to `get_token(cmdline_args.token)`. Should be saver and also more reusable than accessing the one from `__main__`

File: rasa_nlu/server.py [hunk 5]
```diff
@@ -1,7 +1,9 @@
 import argparse
 import logging
-import simplejson
+import os
 from functools import wraps
+
```
Comment (wochinge) [inline rasa_nlu/server.py]:
```suggestion
```

File: CHANGELOG.rst [hunk 6]
```diff
@@ -9,6 +9,7 @@ This project adheres to `Semantic Versioning`_ starting with version 0.7.0.
 
 Added
 -----
+- Authentication token can be also set with env variable `RASA_NLU_TOKEN`.
```
Comment (wochinge) [inline CHANGELOG.rst]: Should we also add this to the documentation?

File: rasa_nlu/server.py (hunk 1 above)
Comment (EPedrotti) [reply rasa_nlu/server.py]: This is very tricky to debug on openshift/kubernets, also if containers have different tokens for NLU the platform won't work properly. I would rather fail early with this than debug the deployment files

Comment (wochinge) [reply rasa_nlu/server.py]: But don't you see it in the container logs? I mean you have to look there anyway if there is an authentication problem right? And if the there is an exception then the pod will be stopped and the logs are not there anymore

Comment (EPedrotti) [reply rasa_nlu/server.py]: The problem is that when you start it with a wrong token it seems to work and in the platform you don't have an error that tells you where the problem is, you get a max recursion exception. This way the problem becomes clear immediately

Comment (wochinge) [reply rasa_nlu/server.py]: mhm, can we then maybe at least check that we only throw this exception when both tokens are different?

File: rasa_nlu/server.py [hunk 7]
```diff
@@ -406,6 +408,27 @@ def unload_model(self, request):
             return simplejson.dumps({"error": "{}".format(e)})
 
 
+def get_token():
+    _clitoken = cmdline_args.token
+
+    if "RASA_NLU_TOKEN" not in os.environ.keys():
+        _envtoken = None
+    else:
+        _envtoken = os.environ["RASA_NLU_TOKEN"]
```
Comment (wochinge) [inline rasa_nlu/server.py]: `os.environ.get("RASA_NLU_TOKEN")` (will return None if not in keys)

File: rasa_nlu/server.py (hunk 1 above)
Comment (EPedrotti) [reply rasa_nlu/server.py]: I think it will just allow an ambiguous behaviour and that there is no advantage in setting both at the same time. 
I did this thinking about our customers. I want to give them a very simple rule to follow for this, just set the token from the cli or as env variable, not both. Also this is better for as well when debugging.

Comment (wochinge) [reply rasa_nlu/server.py]: Ok, but then please create an issue and link it to the issue for the next release (if an issue for the next release exists)

File: rasa_nlu/server.py [hunk 8]
```diff
@@ -409,6 +410,22 @@ def unload_model(self, request):
             return simplejson.dumps({"error": "{}".format(e)})
 
 
+def get_token(_clitoken):
+    _envtoken = os.environ.get("RASA_NLU_TOKEN")
+
+    if _clitoken and _envtoken:
+        raise Exception(
+            "RASA_NLU_TOKEN is set both with the -t option,"
+            " with value `{}`, and with and environment variable, "
```
Comment (wochinge) [inline rasa_nlu/server.py]:
```suggestion
            " with value `{}`, and with an environment variable, "
```

File: rasa_nlu/server.py [hunk 9]
```diff
@@ -409,6 +410,22 @@ def unload_model(self, request):
             return simplejson.dumps({"error": "{}".format(e)})
 
 
+def get_token(_clitoken):
```
Comment (wochinge) [inline rasa_nlu/server.py]:
```suggestion
def get_token(_clitoken: Optional[Text]) -> Text:
```

File: tests/base/test_evaluation.py [hunk 10]
```diff
@@ -1,9 +1,18 @@
 # coding=utf-8
 
+import json
```
Comment (wochinge) [inline tests/base/test_evaluation.py]: can you please revert the changes in this file?
